From 61b138adaeaddefe749f421a3b69c67d4a49a8e3 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.com>
Date: Wed, 11 Oct 2023 11:03:51 +0100
Subject: [PATCH] dt: Add overrides for drm framebuffer allocations on Pi5

Adds dtparam overrides to the base Pi5 DT such that vc4,
DSI0, DSI1, or DPI can be requested to be /dev/fb[012].
No override is specified by default, so the order will be
based on probe order (aka semi-random). Any device that
doesn't have an override specified will be placed above
all specified overrides. Having an fb1 or fb2 override but
no fb0 one will result in no console via fbcon.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
---
 arch/arm/boot/dts/bcm2712-rpi-5-b.dts | 12 ++++++++++++
 arch/arm/boot/dts/overlays/README     | 24 ++++++++++++++++++++++++
 2 files changed, 36 insertions(+)

--- a/arch/arm/boot/dts/bcm2712-rpi-5-b.dts
+++ b/arch/arm/boot/dts/bcm2712-rpi-5-b.dts
@@ -824,5 +824,17 @@ spi10_cs_pins: &spi10_cs_gpio1 {};
 		act_led_trigger = <&act_led>, "linux,default-trigger";
 		pwr_led_activelow = <&pwr_led>, "gpios:8";
 		pwr_led_trigger = <&pwr_led>, "linux,default-trigger";
+		drm_fb0_rp1_dsi0 = <&aliases>, "drm_fb0=",&dsi0;
+		drm_fb0_rp1_dsi1 = <&aliases>, "drm_fb0=",&dsi1;
+		drm_fb0_rp1_dpi = <&aliases>, "drm_fb0=",&dpi;
+		drm_fb0_vc4 = <&aliases>, "drm_fb0=",&vc4;
+		drm_fb1_rp1_dsi0 = <&aliases>, "drm_fb1=",&dsi0;
+		drm_fb1_rp1_dsi1 = <&aliases>, "drm_fb1=",&dsi1;
+		drm_fb1_rp1_dpi = <&aliases>, "drm_fb1=",&dpi;
+		drm_fb1_vc4 = <&aliases>, "drm_fb1=",&vc4;
+		drm_fb2_rp1_dsi0 = <&aliases>, "drm_fb2=",&dsi0;
+		drm_fb2_rp1_dsi1 = <&aliases>, "drm_fb2=",&dsi1;
+		drm_fb2_rp1_dpi = <&aliases>, "drm_fb2=",&dpi;
+		drm_fb2_vc4 = <&aliases>, "drm_fb2=",&vc4;
 	};
 };
--- a/arch/arm/boot/dts/overlays/README
+++ b/arch/arm/boot/dts/overlays/README
@@ -173,6 +173,30 @@ Params:
         cooling_fan             Enables the Pi 5 cooling fan (enabled
                                 automatically by the firmware)
 
+        drm_fb0_rp1_dpi         Assign /dev/fb0 to the RP1 DPI output
+
+        drm_fb0_rp1_dsi0        Assign /dev/fb0 to the RP1 DSI0 output
+
+        drm_fb0_rp1_dsi1        Assign /dev/fb0 to the RP1 DSI1 output
+
+        drm_fb0_vc4             Assign /dev/fb0 to the vc4 outputs
+
+        drm_fb1_rp1_dpi         Assign /dev/fb1 to the RP1 DPI output
+
+        drm_fb1_rp1_dsi0        Assign /dev/fb1 to the RP1 DSI0 output
+
+        drm_fb1_rp1_dsi1        Assign /dev/fb1 to the RP1 DSI1 output
+
+        drm_fb1_vc4             Assign /dev/fb1 to the vc4 outputs
+
+        drm_fb2_rp1_dpi         Assign /dev/fb2 to the RP1 DPI output
+
+        drm_fb2_rp1_dsi0        Assign /dev/fb2 to the RP1 DSI0 output
+
+        drm_fb2_rp1_dsi1        Assign /dev/fb2 to the RP1 DSI1 output
+
+        drm_fb2_vc4             Assign /dev/fb2 to the vc4 outputs
+
         eee                     Enable Energy Efficient Ethernet support for
                                 compatible devices (default "on"). See also
                                 "tx_lpi_timer". Pi3B+ only.
